### ISSUE TITLE
NodeJSEnv in 1.x does not tolerate `source-map-support` to be absent anymore

### ISSUE THREAD

**Issue by sjrd** [scala-js/scala-js#3400]
I discovered that while trying to test the satellite libraries with 1.0.0-M4 before publishing them. For example see here a failed job for scalajs-java-time: https://travis-ci.org/scala-js/scala-js-java-time/jobs/398785844#L1354

This seems to have been introduced in 1acffde2e96877bc6d7b280e3dc35ded7242129e, particularly here:
https://github.com/scala-js/scala-js/commit/1acffde2e96877bc6d7b280e3dc35ded7242129e#diff-f111e3e5727bbd6e0ed3d457a32af359R74
which changed compared to
https://github.com/scala-js/scala-js/commit/1acffde2e96877bc6d7b280e3dc35ded7242129e#diff-29501c0410c2a5694836cf35a3e227c6L61
where `require('source-map-support')` was in a `try`.

@gzm0 Given the parent commit 891e58330f6484f6f828c685092daa1316221a7a, this appears to have been "intentional". But it seems that it is a big usability issue.

I'm considering declaring this a blocker issue, and by-pass 1.0.0-M4. We should promptly fix it and republish 1.0.0-M5, pretending M4 did not exist.

WDYT?

**Comment by gzm0** [comment]
Hum, I'd rather think about this as WAI: If you do not want source map support, you need to disable it. The reason I dropped the try is because otherwise there is no other way to make sure source map support is enabled; that's not good.

Before we had the ability to detect it so one could use this if one needed to be sure (and we did in our build).

> But it seems that it is a big usability issue.

What is the usability issue? That one has to explicitly disable source map support if one does not want it / it isn't installed?

**Comment by sjrd** [comment]
Yes that's basically the usability issue. It is particularly annoying for developers of cross-compiling libraries who don't want to install anything JS-related. It was hard enough, for some of them, to convince them that Node.js was needed in the forest place. And those are the people we really don't want to annoy, because it can make the difference between keeping support for Scala.js and dropping it.

> The reason I dropped the try is because otherwise there is no other way to make sure source map support is enabled; that's not good.

Does anyone ever need to do this? There could be other JS envs for which it is impossible to tell anyway. If the user really wants to know, they would have to know which kind of JS env anyway, and then they could perform themselves the test that we did before.

**Comment by sjrd** [comment]
Another point is that, in the same context of the cross-compiling library (but also other projects), needing source map support is not a property of the *project* per se, but of the local environment. It's very possible that many developers on a cross-compiling project never directly work on the JS version--they only work on the JVM version and let the CI validate JS. Those developers wouldn't want to install source-map-support, and neither the CI. But one developer working on JS-specific stuff, or debugging a JS bug, would like to get the benefit of source maps.

So that means that recommending to cross-compiling libraries that they configure `NodeJSEnv` with `.withSourceMap(false)` in their build is not a good idea.

Perhaps there is a middle ground using a triple-state configuration rather than just `true/false`:
```scala
sealed abstract class SourceMaps

object SourceMaps {
  case object Disable extends SourceMaps
  case object EnableIfAvailable extends SourceMaps
  case object Enable extends SourceMaps
}

def withSourceMaps(sourceMaps: SourceMaps) ...

// and for source compat and convenience
def withSourceMaps(sourceMaps: Boolean) =
  withSourceMaps(if (sourceMaps) SourceMaps.Enable else SourceMaps.Disable)
```

**Comment by gzm0** [comment]
> Does anyone ever need to do this?

The Scala.js project does. Otherwise source map tests will fail strangely.

> There could be other JS envs for which it is impossible to tell anyway.

This is a Node.js specific issue.

>  needing source map support is not a property of the project per se, but of the local environment.

That is a very fair point indeed, so probably we should go with tri-state config with the default being `EnableIfAvailable`. It seems that this is the most reasonable thing to do for most projects. (This will also give us a good location to document the trade-offs discussed in this issue).
